train.py

```diff
@@ -6,51 +8,300 @@
+
+        step = np.array(state['step'])
+        if step % self._checkpoint_every_n == 0:
+            path = os.path.join(self._checkpoint_dir,
+                                'checkpoint_{:07d}.pkl'.format(step))
+            checkpoint_state = jax.device_get(state)
+            logging.info('Serializing experiment state to %s', path)
+            with open(path, 'wb') as f:
+                pickle.dump(checkpoint_state, f)
+
+        return state, metrics
+
+
+class Evaluater:
+    """A class to evaluate the model with, save and checkpoint loss metrics."""
+    def __init__(self, net, loss_fn, checkpoint_dir: str,
+            checkpoint_every_n: int, eval_every_n: int):
+        self._net_init = net.init
+        self._net_apply = net.apply
+        self._loss_fn = loss_fn
+        self.val_queue = []
+        self.best_state = None # save the state with lowest validation error in best state
+        self.lowest_val_loss = None
+        self._checkpoint_dir = checkpoint_dir
+        self._checkpoint_every_n = checkpoint_every_n
+        self._eval_every_n = eval_every_n
+        # load loss curve if metrics file exists in checkpoint_dir
+        metrics_path = os.path.join(self._checkpoint_dir, 'metrics.pkl')
+        best_state_path = os.path.join(self._checkpoint_dir, 
+                'best_state.pkl')
+        if os.path.exists(metrics_path):
+            # load metrics, if they have been saved before
+            logging.info('Loading metrics from %s', metrics_path)
+            with open(metrics_path, 'rb') as f:
+                self._metrics_dict = pickle.load(f)
+            self._loaded_metrics = True
+            # load best state and lowest loss
+            with open(best_state_path, 'rb') as f:
+                best_state_dict = pickle.load(f)
+                self.best_state = best_state_dict['state']
+                self.lowest_val_loss = best_state_dict['loss']
+        else:
+            self._loaded_metrics = False
+
+    @functools.partial(jax.jit, static_argnums=0)
+    def _evaluate_step(
+            self, state: dict, graphs: jraph.GraphsTuple) -> float:
+        """Calculate the mean loss for a batch of graphs."""
+        mean_loss = self._loss_fn(state['params'], graphs, self._net_apply)
+        return mean_loss
+
+    def evaluate_split(
+            self, 
+            state: dict,
+            graphs: Sequence[jraph.GraphsTuple],
+            batch_size: int) -> float:
+        """Return mean loss for all graphs in graphs."""
+        
+        reader = DataReader(data=graphs, 
+            batch_size=batch_size, repeat=False)
+        
+        loss_list = [self._evaluate_step(state, batch) for batch in reader]
+        return np.mean(loss_list)
+
+    def evaluate_model(
+            self, 
+            state: Dict,
+            datasets: Dict[str, Iterable[jraph.GraphsTuple]],
+            splits: Iterable[str]) -> Dict[str, float]:
+        """Return mean loss for every split in splits."""
```

Comment:
Intended use case would be good here too.

E.g. normally splits contains an interable with the train/eval/val sets. This function will then iterate over those data splits and return metrics for those splits. It looks like it only returns eval loss though.

Might be good to add an args section e.g.

```

Adding something like this:

```
  Typical usage example:

  foo = ClassFoo()
  bar = foo.FunctionBar()
```
is always good too
def fetch_smalltable_rows(table_handle: smalltable.Table,
                          keys: Sequence[Union[bytes, str]],
                          require_all_keys: bool = False,
) -> Mapping[bytes, Tuple[str, ...]]:
    """Fetches rows from a Smalltable.

    Retrieves rows pertaining to the given keys from the Table instance
    represented by table_handle.  String keys will be UTF-8 encoded.

    Args:
        table_handle: An open smalltable.Table instance.
        keys: A sequence of strings representing the key of each table
          row to fetch.  String keys will be UTF-8 encoded.
        require_all_keys: If True only rows with values set for all keys will be
          returned.

    Returns:
        A dict mapping keys to the corresponding table row data
        fetched. Each row is represented as a tuple of strings. For
        example:

        {b'Serak': ('Rigel VII', 'Preparer'),
         b'Zim': ('Irk', 'Invader'),
         b'Lrrr': ('Omicron Persei 8', 'Emperor')}

        Returned keys are always bytes.  If a key from the keys argument is
        missing from the dictionary, then that row was not found in the
        table (and require_all_keys must have been False).

    Raises:
        IOError: An error occurred accessing the smalltable.
    """
```